canonical/mongodb-operator
# Test storage re use

### Solution

Tests that when a unit `unit B` gets removed (without removing its storage `storage B`) and a new unit `unit C` re-uses `storage B`, that MongoDB correctly re-uses that storage. (ie the mongo daemon (`mongod`) doesn't just copy all the data from the primary to `unit C` but instead only copies the necessary data to `unit C` based on the re-used `storage B`)

### Specialised Knowledge

1. Storage re-attachment cannot be done on rootfs file systems, see [this convo](https://chat.canonical.com/canonical/pl/eirmfogfx3rmufmom9thjx6pwr)
2. python-libjuju does not have functionality for storage based operations see [issue](https://github.com/juju/python-libjuju/issues/695) and [issue](https://github.com/juju/python-libjuju/issues/694)
3. To verify that MongoDB properly re-uses storage log messages for startup are used:
- If member transitions to STARTUP2 from STARTUP then it is syncing/getting data from primary.
- If member transitions to STARTUP2 from REMOVED then it is re-using the storage we provided.

### Release Notes

1. Created storage helper functions in `tests/integration/ha_tests/helpers.py`
2. Tested storage re-use in `tests/integration/ha_tests/test_ha.py`

### Testing

Since GitHub CI for this charm uses lxd for integration testing rootfs storage is used so this test is not run on the GitHub CI, therefore it must be tested somewhere else. To verify the test works:
1. Create an AWS/GCP/OpenStack cloud. Networking for AWS/GCP/OpenStack instances must be enabled.
4. Create a storage pool: `juju create-storage-pool mias-ebs ebs volume-type=standard`
5. Deploy 1 or more units of application with said storage pool: `juju deploy ./*charm --storage mongodb=mias-ebs,7G,1`
6. Run tests in the model where you deployed the app `tox -e ha-integration -- --model=default`
